### DISCUSSION TITLE
Prevent automatic change of cwd when navigating files in a mono repo

### DISCUSSION THREAD

**Discussion by simonvizzini** [LunarVim/LunarVim#2238] [Q&A] [answered]
Hey everyone,

I decided to ditch my own Neovim config because it was too much of a hassle to keep up with the constant changes in the plugin ecosystem and decided to give LunarVim a shot. So far I'm really happy with it but there is one thing that is really annoying me and I can't figure out which option is responsible for it.

In my current project I'm working on a JS project powered by lerna, which allows us to make use of a "mono repo" setup with multiple JS apps and libraries inside a single git repo.

The simplified repo structure is something like this:

````
packages
  app1
    src
      index.ts
    package.json
  app2
    src
      index.ts
    package.json
  lib1
    src
      index.ts
    package.json
  lib2
    src
      index.ts
    package.json
lerna.json
package.json
````

I open LunarVim at the root of the repo, but whenever I navigate to a file inside one of the package folders (either by Nvim-Tree or Telescope file picker) it will change the working directory to the root of the package folder, e.g. if I open `packages/app1/src/index.ts` it will cwd into `packages/app1`. How can I prevent that? I want the working directory to stay at the repo root all the time.

I tried to set the various `cwd` options of Nvim-Tree, like these:

````
lvim.builtin.nvimtree.setup.update_cwd = false
lvim.builtin.nvimtree.setup.update_focused_file.update_cwd = false
````

but to no avail. I checked LunarVims default Telescope config [here](https://github.com/LunarVim/LunarVim/blob/rolling/lua/lvim/core/telescope.lua) but I don't see anything related to cwd there.

I'd appreciate any hints, it's the only thing holding me back from being productive with lunarvim. Thanks a lot :)

**Comment by simonvizzini** [comment, marked as answer]
Finally found the place responsible for changing the cwd. It's on this line: https://github.com/LunarVim/LunarVim/blob/ff9ae8fdc17d06527d7b8a638c551b0a13b734d7/lua/lvim/core/nvimtree.lua#L120

Basically if `lvim.builtin.project.active` is `true` (which is the default) it will set the nvimtree option `update_cwd` also to true, effectively overriding my own config. So disabling this project feature did solve my issue, although now I can't use this feature anymore, which is unfortunate but I can live without it for now.

**Reply by simonvizzini** [reply]
Unfortunately I have to re-open my question because disabling LVim's `project` feature only halfway solved my problem. For example Nvim-Tree will correctly stay at the repository root when I open a file inside one of the packages, or the Telescope file picker will include all files starting from the repo root, but using the "live grep" Telescope picker will only grep files within the package folder that I've last opened a file from. I double checked with the `:pwd` command and indeed my cwd is still getting changed depending on the location of the file that I open.

To illustrate:

On the command line:
````bash
cd ~/projects/my-project
lvim
````

Then I use the `:pwd` command to see the cwd and it will print `~/projects/my-project` as expected

But if I open a file within lvim by any means (either via nvim-tree, filer picker or `:e` command):
````
:e packages/app1/src/main.ts
````

`:pwd` will now print `~/projects/my-project/packages/app1`. How can I prevent that? I grepped the whole LunarVim repo for `cwd`, `autochdir` etc, but I cannot find any option left that might be responsible for automatically changing my cwd. I didn't have this behavior in my own custom Neovim config so it must be either some LunarVim specific line somewhere, or some of the core plugins is responsible for this behavior, but I'm at a loss what it could be.

Would really appreciate any hints how to disable this automatic changing of my cwd, it's the only thing holding me back to fully embrace LunarVim because live grep is one of the most fundamental features I use all the time during my daily work.

**Reply by randomgeek78** [reply]
I was able to disable automatic directory change with the following changes - 

```lua
local nvimtree = lvim.builtin.nvimtree
nvimtree.setup.update_focused_file.update_cwd = false

local project = lvim.builtin.project
project.manual_mode = true
```

**Reply by simonvizzini** [reply]
Thanks! That works!
Edit: can't mark your comment as answer so feel free to post it as direct answer and I'll mark it as answer

**Reply by perkfly** [reply]
This is so annoying, I spent half an hour to make a minimal reproducible directory to verify that `package.json` was the cause. Thanks for the solution, saved my day.

**Reply by instantaphex** [reply]
Is there someway to enable this on a per project basis?  I actually like the functionality but I have some projects that are monorepos and some that aren't.  I would like to keep this functionality for non-monorepo projects.

**Reply by simonvizzini** [reply]
Actually I'm facing the same issue again after the switch to alpha-nvim, or maybe something else is now responsible for automatic switches of the cwd. If someone has an idea I'd really appreciate it, so far I've been unable to figure out what is going on. I completely disabled "alpha" but that didn't help at all.

My main issue is now with prettier setup as formatter via null-ls. Our mono repo as a prettier config at the root of the repo and if the cwd is the repo root everything works fine, prettier will find the config and apply it correctly. But as soon as I navigate to another part of the mono repo prettier suddenly will use default settings for formatting, it seems to no longer use the config file at repo root.

Here are the relevant parts of my LunarVim config:

````
lvim.builtin.alpha.active = false
lvim.builtin.project.active = false
lvim.builtin.nvimtree.setup.update_cwd = false
lvim.builtin.nvimtree.setup.update_focused_file.update_cwd = false

local formatters = require("lvim.lsp.null-ls.formatters")
formatters.setup({
	{
		-- each formatter accepts a list of options identical to https://github.com/jose-elias-alvarez/null-ls.nvim/blob/main/doc/BUILTINS.md#Configuration
		exe = "prettierd",
		---@usage arguments to pass to the formatter
		args = { "--config-precedence", "prefer-file" },
	},
	{ exe = "eslint_d" },
	{ exe = "stylua" },
})
````

Prettier has the [--find-config-path](https://prettier.io/docs/en/cli.html#--find-config-path-and---config) CLI arg to help find the config file, which when manually executed anywhere inside our mono repo folder structure works as expected, and I assume `--config-precedence=prefer-file` uses the same mechanism to find the config, but somehow as soon as I navigate files of the mono repo inside LunarVim it suddenly stops working, so I assume it's again some cwd issue, but not sure.

Again if someone has any hints or ideas I'd be super grateful. I've been postponing this issue for too long and it became just too annoying to ingore any longer :)

**Reply by simonvizzini** [reply]
Turns out my issue was not caused by a changing CWD but due to a bug in the null-ls language server formatter override which is present in the current stable branch. The rolling branch does not have this issue and I have since switched to rolling.

**Reply by gligorot** [reply]
For anyone reading now, `update_cwd` in the meantime has been renamed to `update_root`. [source](https://github.com/nvim-tree/nvim-tree.lua/blob/master/doc/nvim-tree-lua.txt#L525)

Also, the accepted answer doesn't work anymore, `pwd`still changes when navigating in a monorepo project. Any updates on the fixes on your side @simonvizzini @randomgeek78 ?

**Reply by gligorot** [reply]
For me it's more annoying that `Telecope live_grep` uses pwd (and not the directory where lvim was ran from) than nvimtree changing the visible folder, so I'll try to find solutions in that direction. Will report back if I find anything.

**Reply by gligorot** [reply]
Update: this works, but it's obviously very inflexible for multiple projects:
`lvim.keys.normal_mode["<Space>tg"] = "<cmd>Telescope live_grep cwd=/Users/.../Projects/monorepo-project <CR>"`

I tried to find whether `nvim` keeps the starting directory* somewhere, but didn't manage to. I'd expect then it would be easy to say
`Telescope live_grep cwd=$starting_directory` or something similar, but my nvim/Lua knowledge is low so I can't be sure.

\* for example
```
cd Projects/monorepo-project
lvim .
# $starting directory would be /Users/.../Projects/monorepo-project
```

**Reply by gligorot** [reply]
I think I found a better fix for this:
```
-- lvim.builtin.project.detection_methods = { "pattern" } -- lvim default [1]
lvim.builtin.project.patterns = { ">Projects", ".git" } -- defaults include other VCSs, Makefile, package.json
```
The ">Projects" pattern basically says that root can be a direct child of the folder "Projects" pattern.
".git" is set as a backup.

Now when navigating with telescope it works as expected and keeps the root fixed. But when actually changing a project or going to config.lua ([space]Lc) it changes as expected.

In the weird case that it does switch to a project inside the monorepo, running :ProjectRoot seems to always go back in the tree and correctly finds the correct monorepo root.

[[1]](https://github.com/LunarVim/LunarVim/blob/5ae42b0c44c4237912dcea18457582b52d53804e/lua/lvim/core/project.lua)

**Reply by mkarbo** [reply]
MVP both of you

**Comment by danielo515** [comment]
You can also configure how the CWD is decided. So for example, you can say that projects are first defined by the presence of lerna.json, then the project will be based on the presence of that file first, and fallback to git or package.json after that

**Reply by simonvizzini** [reply]
Hey @danielo515, how exactly do you configure that in LunarVim? Is it some plugin config or LVim config itself? For example I know that null-ls has a `root_dir` option to specify the root dir by providing file names to look for, like lerna.json etc. That's how I did it in my own neovim config and it worked fine, but I checked the LunarVim code and I can't see this option being used, so there must be other ways to configure this somehow.

**Reply by danielo515** [reply]
It is a project feature, but can be configured via lvim like this:

```lua
lvim.builtin.project.patterns = { ".git", "package.json" }
```

I don't have much experience with it, but I think it gives precedence to earlier names, so in this case it will use the git folder and fallback to package.json

**Reply by simonvizzini** [reply]
Oh I see, I actually tried that as well as disabling project completely via `lvim.builtin.project.active = false` and in my case it didn't help. Seems that project is not responsible for my issue, it has to be triggered by something else..

**Reply by rebuilt** [reply]
After setting project to false, run `PackerClean` to uninstall the plugin.

**Reply by simonvizzini** [reply]
Turns out my issue was not caused by changing CWD but due to a bug in the null-ls language server formatter override which is present in the current stable branch. The rolling branch does not have this issue and I have since switched to rolling.

**Comment by ubmit** [comment]
Is there an "official" way of handling different roots?

**Comment by nhdtrung** [comment]
This one works for me
```
lvim.builtin.nvimtree.setup.update_focused_file.update_cwd = false
lvim.builtin.project.manual_mode = true
```
If it doesn't work, use can try those commands:
```
lvim.builtin.nvimtree.setup.update_cwd = false
lvim.builtin.alpha.active = false
lvim.builtin.project.active = false
lvim.builtin.nvimtree.setup.update_focused_file.update_cwd = false
lvim.builtin.nvimtree.actions.change_dir.enable=false
```

**Reply by hamedbavafa** [reply]
Thank you for the spot on reply. I believe the first solution should work in all cases with just a minor adjustment.

```
lvim.builtin.nvimtree.setup.update_focused_file.update_root = false
lvim.builtin.project.manual_mode = true
```

**Comment by plaffitt** [comment]
In my case, this worked:

```
lvim.builtin.nvimtree.setup.update_focused_file.update_root = false
```
